monkey/agent_plugins/exploiters/mssql/src/mssql_client.py

```diff
@@ -1,6 +1,44 @@
-class StubMSSQLClient:
-    def __init__(self):
-        self._authenticated = False
+import logging
+from ipaddress import IPv4Address
+from time import sleep
+from typing import Optional, Set
 
-    def connected(self) -> bool:
-        return self._authenticated
+import pymssql
+
+from common.credentials import Credentials
+from common.types import NetworkPort
+
+logger = logging.getLogger(__name__)
+QUERY_BUFFER_TIME = 0.5
+CONNECTION_CHECK_QUERY = "SELECT 1"
+
+
+class MSSQLClient:
+    def __init__(self, server_timeout: float, query_buffer_time: float = QUERY_BUFFER_TIME):
+        self._cursor: Optional[pymssql.Cursor] = None
+        self._server_timeout = server_timeout
+        self._query_buffer_time = query_buffer_time
+
+    def login(self, ip: IPv4Address, ports: Set[NetworkPort], credentials: Credentials):
+        for port in ports:
+            connection = pymssql.connect(
+                str(ip),
+                credentials.identity.username,
+                credentials.secret.password.get_secret_value(),
+                port=port,
+                login_timeout=self._server_timeout,
+                timeout=self._server_timeout,
+            )
+            try:
+                self._cursor = connection.cursor()
+                self._cursor.execute(CONNECTION_CHECK_QUERY)
+                self._cursor.fetchall()
+                break
+            except pymssql.OperationalError:
+                logger.debug("MSSQL connection is not active")
```

Comment:
How does the caller know whether or not login was successful?

Reply:
`SELECT 1` is just a small query that returns 1 if we were able to connect to the mssql server. We check what the server returns using `fetchall`